### PR TITLE
Expand function-like (bang) procedural macros

### PR DESCRIPTION
**Pull request by vlad20012**
Support function-like procedural macros like `foo!()`. The support includes most of the features that work with declarative macros (aka macro by example) - name resolution, highlighting, code completion, etc.

## User guide

This is an experimental feature, hence bugs are expected. The support is limited to function-like procedural macros (like `foo!()`).
If you want to try it, you should enable `org.rust.macros.proc` and `org.rust.cargo.evaluate.build.script` experimental features (`Ctrl` + `Shift` + `A` -> type "experimental features") and then refresh the cargo project (`Ctrl` + `Shift` + `A` -> type "Refresh Cargo Projects"). If a procedural macro is a part of your project workspace, you should refresh the cargo project (manually) every time you change you proc macro code. 

## Historical notes and implementation details

Proc macro support possibilities were initially researched during JetBrains
internship in 2018. At that time we finalized such design:

```
[IDE] ---------> [Proc macro expander (server)] ----> [proc_macro.so]
       ^ some protocol (e.g. JSON based)         ^ dynamic linkage
```

Basic idea is that the Rust compiler produces dynamic libraries (.dll, .so, .dylib) for proc macro crates.
Then we can link such dynamic libraries and request them to expand a macro that we need. In theory.
But in practice, in 2018 these libraries used unstable Rust ABI and a lot of internal compilers
data structures that made this approach rather impossible.

But some time later some groundwork was done in the Rust compiler, and proc macro libraries started exporting
stable C ABI. Also, there was a compact abstraction layer in `libproc_macro`, so using
`#![feature(proc_macro_internals)]` it became possible to link proc macro libraries and call functions from them.

This approach was implemented in this repo: https://github.com/fedochet/rust-proc-macro-expander, but
it hasn't been integrated into intellij-rust because our macro expansion infrastructure was not ready
for proc macros and because we prioritized other directions. 

Then, rust-analyzer developers used this repo as a base for their proc macro expander! They improved it in several ways and
[published it on crate.io](https://crates.io/crates/ra_ap_proc_macro_srv). Now we can just use `ra_ap_proc_macro_srv` crate from crates.io.

So, I've pushed a simple rust program that embeds `ra_ap_proc_macro_srv` in #6658, then @Undin set up its builds on CI in #6713. Then, `proc-macro-expander` was renamed to `intellij-rust-native-helper` in #6735 because the binary started to be responsible for other stuff, not only proc macro expansion.

## TODO
Add tests for completion, `extend selection`, `type info`

### CODE REVIEW

**Comment by vlad20012** [comment]
It's ready for review. I've updated the PR description.

In this PR we only support function-like procedural macros like `foo!()`. The support includes most of the features that work with declarative macros (aka macro by example) - name resolution, highlighting, code completion, etc. If you want to try it, you should enable `org.rust.macros.proc` and `org.rust.cargo.evaluate.build.script` experimental features. Other kinds of procedural macros (custom derive and attribute macros) will be supported in other PRs

**Comment by hamza1311** [comment]
> The support includes most of the features that work with declarative macros (aka macro by example) - name resolution, highlighting, code completion, etc

How does highlighting and code completion work with macros that don't take syntactically correct rust as their input? I don't see anything related to such macros from a quick look at test cases. Are they not supported yet? I'm assuming the support is only for their output, correct? If so, is there any plans to allow adding support custom syntax provided by proc macros (like I described in #6367)?

**Comment by vlad20012** [comment]
@hamza1311, Completion for procedural macros works similar to completion for declarative macros - we expand the macro (if we can) and then invoke the completion on the macro expansion (with caret position mapping). This is expected to work worse for procedural macros because they are (usually) not resilient to incorrect rust syntax (this should be solved on the procedural macro side, i.e. in `syn` crate)
There are ideas on how we could support very custom syntax (again, this should be done on the proc macro side), see https://github.com/rust-analyzer/rust-analyzer/issues/7402#issuecomment-770196608

I can't test `yew` for now because `html` macro is defined using `proc_macro_hack` that is an attribute macro which are not covered by this PR, but I'd expect a lot of code completion working there, at least in lambdas (because they are just arbitrary rust code) and in custom component names (because they are just a rust struct names)

#6367 is a valid issue anyway

**Comment by hamza1311** [comment]
> we expand the macro (if we can) and then invoke the completion on the macro expansion (with caret position mapping).

I'm not sure how much of this is done on IntelliJ's side so correct me if I'm wrong but would it be possible to provide a way to set the environment where the expanded code is invoked? For example, the [`sqlx::query!`](https://docs.rs/sqlx/0.5.1/sqlx/macro.query.html) family of macros make calls to database so I think ability to force offline mode (where it reads from a pre-generated file instead of calling the database) in those macros is important.

> I can't test yew for now because html macro is defined using proc_macro_hack that is an attribute macro which are not covered by this PR, but I'd expect a lot of code completion working there in the future, at least in lambdas (because they are just arbitrary rust code) and in custom component names (because they are just a rust struct names)

`Yew` doesn't use `proc_macro_hack` anymore. That change isn't released so if you wish to test that macro, you can depend from master. You can find the macro on master [here](https://github.com/yewstack/yew/blob/master/packages/yew-macro/src/lib.rs#L112-L116).

**Comment by vlad20012** [comment]
> You can find the macro on master here

Oh, thanks! Yep, the completion works as expected:

https://user-images.githubusercontent.com/3221931/107040448-49908a80-67d0-11eb-8137-a214d44547e6.mp4

> I'm not sure how much of this is done on IntelliJ's side so correct me if I'm wrong but would it be possible to provide a way to set the environment where the expanded code is invoked? For example, the sqlx::query! family of macros make calls to database so I think ability to force offline mode (where it reads from a pre-generated file instead of calling the database) in those macros is important.

We could tell a proc macro that it is launched from IDE (In this PR I set env var `INTELLIJ_RUST=true`), but a proc macro should take it into account.

**Comment by hamza1311** [comment]
> Yep, the completion works as expected:

Incredible! Looking forward to the release.

File: src/main/kotlin/org/rust/lang/core/macros/decl/DeclMacroExpander.kt [hunk 1]
```diff
@@ -0,0 +1,518 @@
+/*
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/macros/decl/DeclMacroExpander.kt:1]: Maybe split files move into separate commit

**Review by dima74 — APPROVED** [review]
🚀🚀🚀

File: src/main/kotlin/org/rust/lang/core/macros/RangeMap.kt [hunk 2]
```diff
@@ -122,3 +123,20 @@ private fun MappedTextRange.dstIntersection(range: TextRange): MappedTextRange?
         null
     }
 }
+
+fun MutableList<MappedTextRange>.mergeAdd(range: MappedTextRange) {
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/macros/RangeMap.kt:127]: Actually, I don't fully understand why we need this. Could you explain, please (probably, it's worth adding a doc comment with an example)?

File: src/main/kotlin/org/rust/ide/wordSelection/RsMacroCallSelectionHandler.kt [hunk 3]
```diff
@@ -54,15 +54,19 @@ class RsMacroCallSelectionHandler : ExtendWordSelectionHandlerBase() {
         val expansion = macroCall.expansion ?: return null // impossible?
         val expansionText = expansion.file.text
 
+        if (offsetInExpansion > expansionText.length) return null
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/wordSelection/RsMacroCallSelectionHandler.kt:57]: Is it really possible?

**Review by Undin — CHANGES_REQUESTED** [review]
In general, looks awesome! 🚀 

It would be great to write more tests for existing features like completion, highlighting, extend  selection, etc. but I don't insist to do it in this PR - it's already quite huge

Also, some tests fail with these changes

File: src/main/kotlin/org/rust/lang/core/macros/proc/Request.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package org.rust.lang.core.macros.proc
+
+import com.google.gson.*
+import com.google.gson.annotations.SerializedName
+import org.rust.lang.core.macros.tt.TokenTree
+import java.lang.reflect.Type
+
+sealed class Request {
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/macros/proc/Request.kt:13]: There is a `ListMacros` request in the protocol. I'm not going to use it, but I'd want to specify that there are other request kinds

File: src/main/kotlin/org/rust/lang/core/macros/RangeMap.kt (hunk 2 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/macros/RangeMap.kt:127]: It's an optimization to reduce the list size. The method replaces the last list element with a united range if they intersect

**Review by dima74 — APPROVED** [review]

File: src/main/kotlin/org/rust/lang/core/macros/RsMacroDataWithHash.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.rust.lang.core.psi.RsFunction
+import org.rust.lang.core.psi.RsMacro
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve2.DeclMacroDefInfo
+import org.rust.lang.core.resolve2.MacroDefInfo
+import org.rust.lang.core.resolve2.ProcMacroDefInfo
+import org.rust.stdext.HashCode
+
+class RsMacroDataWithHash<T: RsMacroData>(
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/macros/RsMacroDataWithHash.kt:16]:
```suggestion
class RsMacroDataWithHash<T : RsMacroData>(
```

File: src/main/kotlin/org/rust/lang/core/macros/RsMacroDataWithHash.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    val data: T,
+    val bodyHash: HashCode?
+) {
+    fun mixHash(call: RsMacroCallDataWithHash): HashCode? {
+        val callHash = when (data as RsMacroData) {
```
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/macros/RsMacroDataWithHash.kt:21]: Actually, it's a kotlin plugin false positive. Let's just add a suppression here

**Review by Undin — COMMENTED** [review]
* The current implementation cannot expand https://github.com/yewstack/yew/blob/f93fa3c8b2106dca4eb768ae005ef530310b7721/examples/boids/src/main.rs#L118-L166
* expand selection can't select `Msg::RestartSimulation` in https://github.com/yewstack/yew/blob/f93fa3c8b2106dca4eb768ae005ef530310b7721/examples/boids/src/main.rs#L92
* Type info doesn't work any expr in https://github.com/yewstack/yew/blob/f93fa3c8b2106dca4eb768ae005ef530310b7721/examples/boids/src/main.rs#L92

File: src/main/kotlin/org/rust/lang/core/macros/RsMacroCallDataWithHash.kt [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    val bodyHash: HashCode?
+) {
+    fun hashWithEnv(): HashCode? {
+        val bodyHash = bodyHash ?: return null
+        val env = data.packageEnv
+            .entries
+            .toList()
+            .sortedBy { it.key }
+            .joinToString(separator = ";") { it.key + "=" + it.value }
+        val envHash = HashCode.compute(env)
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/macros/RsMacroCallDataWithHash.kt:21]: I don't think it's ok, but let's please fix it later :)

**Comment by vlad20012** [comment]
> The current implementation cannot expand https://github.com/yewstack/yew/blob/f93fa3c8b2106dca4eb768ae005ef530310b7721/examples/boids/src/main.rs#L118-L166

Fixed. It was a bug in our TokenTree parser.

> expand selection can't select Msg::RestartSimulation in https://github.com/yewstack/yew/blob/f93fa3c8b2106dca4eb768ae005ef530310b7721/examples/boids/src/main.rs#L92
Type info doesn't work for any expr in https://github.com/yewstack/yew/blob/f93fa3c8b2106dca4eb768ae005ef530310b7721/examples/boids/src/main.rs#L92

I'll fix it in another PR. The same issues can be also applied to (some specific) declarative macros, I guess. I'll tweak Extend Selection / Type Info later

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansionError.kt [hunk 8]
```diff
@@ -8,44 +8,88 @@ package org.rust.lang.core.macros
...
-    class Nesting(macroCallBody: PsiBuilder, val variableName: String): MacroMatchingError(macroCallBody) {
+
+    class Nesting(macroCallBody: PsiBuilder, val variableName: String) : MacroMatchingError(macroCallBody) {
         override fun toString(): String = "Nesting($variableName)"
     }
 }
+
+sealed class ProcMacroExpansionError : MacroExpansionError() {
+    data class ServerSideError(val message: String) : ProcMacroExpansionError() {
+        override fun toString(): String = "ProcMacroExpansionError.Expansion(message = \"$message\")"
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpansionError.kt:69]:
```suggestion
        override fun toString(): String = "ProcMacroExpansionError.ServerSideError(message = \"$message\")"
```

Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/macros/MacroExpansionError.kt:69]: nit: Do you actually need `ProcMacroExpansionError.` prefix? If not, you can use `javaClass.simpleName` not to forget updating `toString`. If you need it, probably we can use something like `.javaClass.canonicalName.removePrefix("${javaClass.packageName}.")`.

Note, in fact you use `simpleName` for `DeclMacroExpansionError` errors.
Moreover, you can provide basic implementation of `toString` in this way and avoid `override fun toString(): String = "ProcMacroExpansionError.Timeout"` and friends

File: src/test/kotlin/org/rust/lang/core/macros/proc/RsProcMacroExpansionTest.kt [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        }
+        val pkg = project.cargoProjects.singleWorkspace().packages
+            .find { it.name == "my_proc_macro" }!!
+        val lib = pkg.procMacroArtifact?.path?.toString()
+            ?: error("Procedural macro artifact is not found. This most likely means a compilation failure")
+        val server = ProcMacroServerPool.tryCreate(testRootDisposable)
+            ?: error("native-helper is not available")
+        val expander = ProcMacroExpander(project, server)
+
+        expander.apply {
```
Comment (Undin) [inline src/test/kotlin/org/rust/lang/core/macros/proc/RsProcMacroExpansionTest.kt:120]: nit:
```suggestion
        with(expander) {
```
looks `with` is a little bit more suitable here

**Review by Undin — APPROVED** [review]
LGTM except minor comments

**Comment by vlad20012** [comment]
bors r=undin
